### PR TITLE
Feature/add time manager

### PR DESCRIPTION
**Pull request by Renardjojo**
Create a time system according to the diagram about this subject. For now, time system is not a singleton and allow the user to create timer task. This task can loop and can only be destroyed if the container is clear with the function clearScaledTimer or clearUnscaledTimer. If this feature is really useful, I will implement it

### CODE REVIEW

**Review by Korotsu — APPROVED** [review]

File: engine/include/Engine/Core/Rendering/Window/WindowGLFW.hpp [hunk 1]
```diff
@@ -53,7 +53,7 @@ class Window
      * @param height
      * @return
      */
-    void getSize(int& width, int& height) const noexcept;
+    void getSize(int width, int height) const noexcept;
```
Comment (badsami) [inline engine/include/Engine/Core/Rendering/Window/WindowGLFW.hpp]: The new version can't return the size of the window without pointers, references, or a return value
```suggestion
    void getSize(int& width, int& height) const noexcept;
    // Or
    Math::Vec2 getSize() const noexcept;
```

Comment (ThomasProg) [inline engine/include/Engine/Core/Rendering/Window/WindowGLFW.hpp]: Why ?

File: engine/src/WindowGLFW.cpp [hunk 2]
```diff
@@ -53,7 +53,7 @@ Window::~Window() noexcept
     Log::log("GLFW window release");
 }
 
-void Engine::Core::Renderering::Window::getSize(unsigned short& width, unsigned short& height) const noexcept
+void Engine::Core::Renderering::Window::getSize(int width, int height) const noexcept
```
Comment (ThomasProg) [inline engine/src/WindowGLFW.cpp]: Same as before : add references

File: engine/include/Engine/Core/TimeSystem/TimeSystem.hpp [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+#include <functional> //std::function
+#include <queue>
+
+namespace Engine::Core
+{
+struct TimerTask
+{
+    double localTimer = 0.; // if current time egal 1s and local timer egal 0.5 global time egal 1.5
+    double globalTimer = 0.;
+    std::function<void()> task = nullptr;
```
Comment (badsami) [inline engine/include/Engine/Core/TimeSystem/TimeSystem.hpp:20]: **TODO for later**, when `Task` will be implemented: replace `std::function<void()>` with `Task`, so functions can be stored with any number and type of argument

File: engine/include/Engine/Core/TimeSystem/TimeSystem.hpp [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    std::chrono::system_clock::time_point m_tempTime = m_time;
+    double m_fixedTimeAcc = 0.;
+
+    unsigned int m_frameAcc = 0;
+
+    double m_deltaTime = 0.;
+    double m_unscaledDeltaTime = 0.;
+    double m_timeScale = 1.;
+    double m_fixedUnscaledDetlaTime = 1. / 60.;
+    double m_fixedDetlaTime = m_fixedUnscaledDetlaTime * m_timeScale;
```
Comment (badsami) [inline engine/include/Engine/Core/TimeSystem/TimeSystem.hpp]:
```suggestion
    double m_fixedUnscaledDeltaTime = 1. / 60.;
    double m_fixedDeltaTime = m_fixedUnscaledDeltaTime * m_timeScale;
```

File: engine/include/Engine/Core/TimeSystem/TimeSystem.hpp [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+public:
+    /**
+     * @brief Update the time, update system and renderSystem. The update function will call with accumulator to be call
+     * a fixed time by second
+     *
+     * @param fixedUpdateFunction : lambda that contain all fixed update function
+     * @param updateFunction : lambda that contain all update function
+     * @param renderFunction : lambda that contain all render function
+     */
+    void update(std::function<void(double fixedUnscaledDetlaTime, double fixedDetlaTime)> fixedUpdateFunction,
```
Comment (badsami) [inline engine/include/Engine/Core/TimeSystem/TimeSystem.hpp]:
```suggestion
    void update(std::function<void(double fixedUnscaledDeltaTime, double fixedDeltaTime)> fixedUpdateFunction,
```

**Review by ThomasProg — COMMENTED** [review]
getSize() has wrong arguments
Otherwise, valid commit.

**Review by ThomasProg — APPROVED** [review]

File: engine/include/Engine/Core/TimeSystem/TimeSystem.hpp [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+                std::function<void(double unscaledDeltaTime, double deltaTime)> updateFunction,
+                std::function<void()> renderFunction) noexcept;
+
+
+    /**
+     * @brief Change the fixed delta time used in fixed update. By default this value egales : 1/60 or 60 Hz
+     * @param newFixedUnscaledDetlaTime
```
Comment (badsami) [inline engine/include/Engine/Core/TimeSystem/TimeSystem.hpp]:
```suggestion
     * @param newFixedUnscaledDeltaTime 
```

File: engine/include/Engine/Core/TimeSystem/TimeSystem.hpp [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+                std::function<void(double unscaledDeltaTime, double deltaTime)> updateFunction,
+                std::function<void()> renderFunction) noexcept;
+
+
+    /**
+     * @brief Change the fixed delta time used in fixed update. By default this value egales : 1/60 or 60 Hz
+     * @param newFixedUnscaledDetlaTime 
+     * @return 
+    */
+    inline void setFixedUnscaledDeltaTime(double newFixedUnscaledDetlaTime) noexcept;
```
Comment (badsami) [inline engine/include/Engine/Core/TimeSystem/TimeSystem.hpp]:
```suggestion
    inline constexpr void setFixedUnscaledDeltaTime(double newFixedUnscaledDeltaTime) noexcept;
```

File: engine/include/Engine/Core/TimeSystem/TimeSystem.hpp [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    [[nodiscard]] inline double getFixedDeltaTime() noexcept;
+    [[nodiscard]] inline double getFixedUnscaledDeltaTime() noexcept;
+    [[nodiscard]] inline double getDeltaTime() noexcept;
+    [[nodiscard]] inline double getUnscaledDeltaTime() noexcept;
+    [[nodiscard]] inline double getAccumulateTime() noexcept;
+    [[nodiscard]] inline double getAccumulateUnscaledTime() noexcept;
+    [[nodiscard]] inline double getTimeScale() noexcept;
+
+    /**
+     * @brief Redefine time scale. This value will use to comput scaled detla time and scaled fixed delta time. Usefull to change the game speed
```
Comment (badsami) [inline engine/include/Engine/Core/TimeSystem/TimeSystem.hpp]:
```suggestion
     * @brief Redefine time scale. This value will use to compute scaled delta time and scaled fixed delta time. Useful to change the game speed
```

File: engine/src/TimeSystem.cpp [hunk 9]
```diff
@@ -0,0 +1,67 @@
+#include "Engine/Core/TimeSystem/TimeSystem.hpp"
+#include "Engine/Core/Debug/Log.hpp"
+#include <string>
+#include <thread> // std::this_thread::sleep_for
+
+using namespace Engine::Core;
+using namespace Engine::Core::Debug;
+
+void TimeSystem::update(std::function<void(double fixedUnscaledDetlaTime, double fixedDetlaTim)> fixedUpdateFunction,
```
Comment (badsami) [inline engine/src/TimeSystem.cpp]:
```suggestion
void TimeSystem::update(std::function<void(double fixedUnscaledDeltaTime, double fixedDeltaTime)> fixedUpdateFunction,
```

File: engine/src/TimeSystem.cpp [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    m_scaledTimeAcc += m_deltaTime;
+    m_unscaledTimeAcc += m_unscaledDeltaTime;
+
+    m_scaledTimeAcc *= !isinf(m_scaledTimeAcc);     // reset if isInf (avoid conditionnal jump)
+    m_unscaledTimeAcc *= !isinf(m_unscaledTimeAcc); // reset if isInf (avoid conditionnal jump)
+
+    /*Fixed update*/
+    m_fixedTimeAcc += m_unscaledDeltaTime;
+
+    while (m_fixedTimeAcc >= m_fixedUnscaledDetlaTime)
```
Comment (badsami) [inline engine/src/TimeSystem.cpp]:
```suggestion
    while (m_fixedTimeAcc >= m_fixedUnscaledDetlaTime)
```

File: engine/src/TimeSystem.cpp [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        fixedUpdateFunction(m_fixedUnscaledDetlaTime, m_fixedDetlaTime);
```
Comment (badsami) [inline engine/src/TimeSystem.cpp]:
```suggestion
        fixedUpdateFunction(m_fixedUnscaledDetlaTime, m_fixedDeltaTime);
```

File: engine/src/main.cpp [hunk 12]
```diff
@@ -35,15 +43,55 @@ void logExample()
...
+            std::cout << "FPS (fixedUpdate): " << fixedUpdateFrameCount / m_FPSLogDelay << std::endl;
+            std::cout << "FPS (unFixedUpdate): " << unFixedUpdateFrameCount / m_FPSLogDelay << std::endl << std::endl;
+            fixedUpdateFrameCount = 0;
+            unFixedUpdateFrameCount = 0;
+
+     }, true);
+
+    while (1)
+    {
+        ts.update([&](double fixedUnscaledDetlaTime, double fixedDetlaTim)
```
Comment (badsami) [inline engine/src/main.cpp]:
```suggestion
        ts.update([&](double fixedUnscaledDeltaTime, double fixedDeltaTime)
```

File: engine/src/TimeSystem.cpp [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+        m_fixedTimeAcc -= m_fixedUnscaledDetlaTime;
```
Comment (badsami) [inline engine/src/TimeSystem.cpp]:
```suggestion
        m_fixedTimeAcc -= m_fixedUnscaledDeltaTime;
```

File: engine/src/TimeSystem.inl [hunk 14]
```diff
@@ -0,0 +1,72 @@
+#include "Engine/Core/TimeSystem/TimeSystem.hpp"
+
+inline double Engine::Core::TimeSystem::getFixedDeltaTime() noexcept
+{
+    return m_fixedDetlaTime;
```
Comment (badsami) [inline engine/src/TimeSystem.inl]:
```suggestion
    return m_fixedDeltaTime;
```

File: engine/src/TimeSystem.inl [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+inline double Engine::Core::TimeSystem::getFixedUnscaledDeltaTime() noexcept
+{
+    return m_fixedUnscaledDetlaTime;
```
Comment (badsami) [inline engine/src/TimeSystem.inl]:
```suggestion
    return m_fixedUnscaledDeltaTime;
```

File: engine/src/TimeSystem.inl [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+inline void Engine::Core::TimeSystem::setFixedUnscaledDeltaTime(double newFixedUnscaledDetlaTime) noexcept
```
Comment (badsami) [inline engine/src/TimeSystem.inl]:
```suggestion
inline constexpr void Engine::Core::TimeSystem::setFixedUnscaledDeltaTime(double newFixedUnscaledDeltaTime) noexcept
```

File: engine/src/TimeSystem.inl [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+{
+    if (newFixedUnscaledDetlaTime > 1.)
+        Engine::Core::Debug::Log::logWarning("New fixed unscaled detla time > 1.");
```
Comment (badsami) [inline engine/src/TimeSystem.inl]:
```suggestion
    if (newFixedUnscaledDeltaTime > 1.)
        Engine::Core::Debug::Log::logWarning("New fixed unscaled delta time > 1.");
```

File: engine/src/TimeSystem.inl [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+
+    m_fixedUnscaledDetlaTime = newFixedUnscaledDetlaTime;
```
Comment (badsami) [inline engine/src/TimeSystem.inl]:
```suggestion
    m_fixedUnscaledDeltaTime = newFixedUnscaledDeltaTime;
```

File: engine/src/TimeSystem.inl [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+
+inline double Engine::Core::TimeSystem::getTimeScale() noexcept
+{
+    return m_timeScale;
+}
+
+inline void Engine::Core::TimeSystem::setTimeScale(double newtimeScale) noexcept
+{
+    m_timeScale = newtimeScale;
+    m_fixedDetlaTime = m_fixedUnscaledDetlaTime * m_timeScale;
```
Comment (badsami) [inline engine/src/TimeSystem.inl]:
```suggestion
    m_fixedDeltaTime = m_fixedUnscaledDeltaTime * m_timeScale;
```

File: engine/include/Engine/Core/TimeSystem/TimeSystem.hpp [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+    {
+        return globalTimer > other.globalTimer;
+    }
+};
+
+class TimeSystem
+{
+protected:
+    std::chrono::system_clock::time_point m_time = std::chrono::system_clock::now();
+    std::chrono::system_clock::time_point m_tempTime = m_time;
```
Comment (badsami) [inline engine/include/Engine/Core/TimeSystem/TimeSystem.hpp]: According to cppreference, the time returned by `std::chrono::system_clock` can be altered, which can result in a return in time. See cppreference page [here](https://en.cppreference.com/w/cpp/chrono/system_clock#mw-content-text).

I suggest we use `std::chrono::steady_clock`, which is steady (unchanging) and monotonic (time calculated from a fixed point in time). See description [here](https://en.cppreference.com/w/cpp/chrono/steady_clock)

File: engine/include/Engine/Core/TimeSystem/TimeSystem.hpp [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+    */
+    inline void setFixedUnscaledDeltaTime(double newFixedUnscaledDetlaTime) noexcept;
+
+    [[nodiscard]] inline double getFixedDeltaTime() noexcept;
+    [[nodiscard]] inline double getFixedUnscaledDeltaTime() noexcept;
+    [[nodiscard]] inline double getDeltaTime() noexcept;
+    [[nodiscard]] inline double getUnscaledDeltaTime() noexcept;
+    [[nodiscard]] inline double getAccumulateTime() noexcept;
+    [[nodiscard]] inline double getAccumulateUnscaledTime() noexcept;
+    [[nodiscard]] inline double getTimeScale() noexcept;
```
Comment (badsami) [inline engine/include/Engine/Core/TimeSystem/TimeSystem.hpp]:
```suggestion
    [[nodiscard]] inline constexpr double getFixedDeltaTime() const noexcept;
    [[nodiscard]] inline constexpr double getFixedUnscaledDeltaTime() const noexcept;
    [[nodiscard]] inline constexpr double getDeltaTime() const noexcept;
    [[nodiscard]] inline constexpr double getUnscaledDeltaTime() const noexcept;
    [[nodiscard]] inline constexpr double getAccumulatedTime() const noexcept;
    [[nodiscard]] inline constexpr double getAccumulatedUnscaledTime() const noexcept;
    [[nodiscard]] inline constexpr double getTimeScale() const noexcept;
```

File: engine/src/TimeSystem.inl [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+
+inline void Engine::Core::TimeSystem::setFixedUnscaledDeltaTime(double newFixedUnscaledDetlaTime) noexcept
+{
+    if (newFixedUnscaledDetlaTime > 1.)
+        Engine::Core::Debug::Log::logWarning("New fixed unscaled detla time > 1.");
+
+    m_fixedUnscaledDetlaTime = newFixedUnscaledDetlaTime;
+}
+
+inline double Engine::Core::TimeSystem::getDeltaTime() noexcept
```
Comment (badsami) [inline engine/src/TimeSystem.inl]:
```suggestion
inline constexpr double Engine::Core::TimeSystem::getDeltaTime() const noexcept
```

File: engine/src/TimeSystem.inl [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+{
+    return m_deltaTime;
+}
+
+inline double Engine::Core::TimeSystem::getUnscaledDeltaTime() noexcept
```
Comment (badsami) [inline engine/src/TimeSystem.inl]:
```suggestion
inline constexpr double Engine::Core::TimeSystem::getUnscaledDeltaTime() const noexcept
```

File: engine/src/TimeSystem.inl [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+{
+    return m_unscaledDeltaTime;
+}
+
+inline double Engine::Core::TimeSystem::getAccumulateTime() noexcept
```
Comment (badsami) [inline engine/src/TimeSystem.inl]:
```suggestion
inline constexpr double Engine::Core::TimeSystem::getAccumulatedTime() const noexcept
```

File: engine/src/TimeSystem.inl [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+{
+    return m_scaledTimeAcc;
+}
+
+inline double Engine::Core::TimeSystem::getAccumulateUnscaledTime() noexcept
```
Comment (badsami) [inline engine/src/TimeSystem.inl]:
```suggestion
inline constexpr double Engine::Core::TimeSystem::getAccumulatedUnscaledTime() const noexcept
```

File: engine/src/TimeSystem.inl [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+{
+    return m_unscaledTimeAcc;
+}
+
+inline double Engine::Core::TimeSystem::getTimeScale() noexcept
```
Comment (badsami) [inline engine/src/TimeSystem.inl]:
```suggestion
inline constexpr double Engine::Core::TimeSystem::getTimeScale() const noexcept
```

File: engine/include/Engine/Core/TimeSystem/TimeSystem.hpp [hunk 27]
```diff
@@ -0,0 +1,122 @@
...
+
+    /**
+     * @brief Redefine time scale. This value will use to comput scaled detla time and scaled fixed delta time. Usefull to change the game speed
+     * @param newtimeScale 
+     * @return 
+    */
+    inline void setTimeScale(double newtimeScale) noexcept;
```
Comment (badsami) [inline engine/include/Engine/Core/TimeSystem/TimeSystem.hpp]:
```suggestion
    inline constexpr void setTimeScale(double newtimeScale) noexcept;
```

File: engine/src/TimeSystem.inl [hunk 28]
```diff
@@ -0,0 +1,72 @@
...
+{
+    return m_timeScale;
+}
+
+inline void Engine::Core::TimeSystem::setTimeScale(double newtimeScale) noexcept
```
Comment (badsami) [inline engine/src/TimeSystem.inl]:
```suggestion
inline constexpr void Engine::Core::TimeSystem::setTimeScale(double newtimeScale) noexcept
```

File: engine/src/TimeSystem.cpp [hunk 29]
```diff
@@ -0,0 +1,67 @@
...
+    }
```
Comment (badsami) [inline engine/src/TimeSystem.cpp]: I'm not sure I fully understand what this is doing, I'm sorry if I get this wrong: `m_fixedTimeAcc` is incremented by `m_unscaledDeltaTime` every time `TimeSystem::update()` is called, but it is then compared to `m_fixedUnscaledDetlaTime`. Shouldn't it be compared to `m_fixedDeltaTime` instead?

File: engine/include/Engine/Core/TimeSystem/TimeSystem.hpp [hunk 30]
```diff
@@ -0,0 +1,122 @@
...
+};
+
+class TimeSystem
+{
+protected:
+    std::chrono::system_clock::time_point m_time = std::chrono::system_clock::now();
+    std::chrono::system_clock::time_point m_tempTime = m_time;
+    double m_fixedTimeAcc = 0.;
+
+    unsigned int m_frameAcc = 0;
```
Comment (badsami) [inline engine/include/Engine/Core/TimeSystem/TimeSystem.hpp]: `m_frameAcc` is not used anywhere.
If it is meant to be used, move it after the `double`s to align it properly

**Review by badsami — CHANGES_REQUESTED** [review]

**Review by badsami — APPROVED** [review]

**Comment by Renardjojo** [comment]
Done

**Pull request merged**